File: backend/src/main/java/com/NBE3_4_2_Team4/global/security/accessDeniedHandler/CustomAccessDeniedHandler.java
package com.NBE3_4_2_Team4.global.security.accessDeniedHandler;

import com.NBE3_4_2_Team4.global.rsData.RsData;
import com.fasterxml.jackson.databind.ObjectMapper;
import jakarta.servlet.ServletException;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import lombok.RequiredArgsConstructor;
import org.springframework.security.access.AccessDeniedException;
import org.springframework.security.web.access.AccessDeniedHandler;
import org.springframework.stereotype.Component;

import java.io.IOException;
import java.nio.charset.StandardCharsets;

@Component
@RequiredArgsConstructor
public class CustomAccessDeniedHandler implements AccessDeniedHandler {
    private final ObjectMapper objectMapper;

    @Override
    public void handle(HttpServletRequest request, HttpServletResponse response, AccessDeniedException accessDeniedException) throws IOException, ServletException {
        // RsData 형식의 JSON 응답 생성
        RsData<String> responseData = new RsData<>("403-1", "Forbidden", "권한이 부족합니다.");

        // 응답 설정
        response.setContentType("application/json");
        response.setCharacterEncoding(StandardCharsets.UTF_8.name());
        response.setStatus(HttpServletResponse.SC_FORBIDDEN);

        // JSON 변환 후 응답에 작성
        String jsonResponse = objectMapper.writeValueAsString(responseData);
        response.getWriter().write(jsonResponse);
    }
}

File: backend/src/main/java/com/NBE3_4_2_Team4/global/security/accessDeniedHandler/CustomAccessDeniedHandler.kt
package com.NBE3_4_2_Team4.global.security.accessDeniedHandler

import com.NBE3_4_2_Team4.global.rsData.RsData
import com.fasterxml.jackson.databind.ObjectMapper
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.springframework.http.MediaType
import org.springframework.security.web.access.AccessDeniedHandler
import org.springframework.stereotype.Component
import org.springframework.security.access.AccessDeniedException

@Component
class CustomAccessDeniedHandler(
    private val objectMapper: ObjectMapper
) : AccessDeniedHandler {

    override fun handle(
        request: HttpServletRequest,
        response: HttpServletResponse,
        accessDeniedException: AccessDeniedException
    ){
        // RsData 형식의 JSON 응답 생성
        val responseData = RsData("403-1", "Forbidden", "권한이 부족합니다.")


        // 응답 설정
        response.contentType = MediaType.APPLICATION_JSON_VALUE
        response.status = HttpServletResponse.SC_FORBIDDEN


        // JSON 변환 후 응답에 작성
        val jsonResponse = objectMapper.writeValueAsString(responseData)
        response.writer.write(jsonResponse)
    }
}